apps/api-documenter/src/documenters/YamlDocumenter.ts

```diff
@@ -105,67 +118,94 @@ export class YamlDocumenter {
   }
 
   private _visitApiItems(apiItem: ApiDocumentedItem, parentYamlFile: IYamlApiFile | undefined): boolean {
-    const yamlItem: IYamlItem | undefined = this._generateYamlItem(apiItem);
-    if (!yamlItem) {
-      return false;
-    }
-
-    this.onCustomizeYamlItem(yamlItem);
+    const savedKnownTypeParameters: Set<string> | undefined = this._knownTypeParameters;
+    try {
+      // Track type parameters declared by a declaration so that we do not resolve them
+      // when looking up types in _linkToUidIfPossible()
```

Comment:
In the test, this input:
```ts
export interface IDocInterface6 {
  . . .
  genericReferenceMethod<T>(x: T): T;
}
```

...gets emitted as this YAML:
```yaml
  - uid: 'api-documenter-test!IDocInterface6#genericReferenceMethod:member(1)'
    name: genericReferenceMethod(x)
    fullName: genericReferenceMethod(x)
    langs:
      - typeScript
    type: method
    syntax:
      content: 'genericReferenceMethod<T>(x: T): T;'
      return:
        type:
          - T
        description: ''
      parameters:
        - id: x
          description: ''
          type:
            - T
      typeParameters:
        - id: T
```
...with a reference like this:
```yaml
references:
  - uid: T
```

If we really want to treat `T` as a referencable thing, then it probably should get a unique UID notation (e.g. something like `api-documenter-test!IDocInterface6#genericReferenceMethod:member(1).T`).  The `this._knownTypeParameters` lookup here appears to be a workaround for the problem that the name `T` is used for many different things, but is being treated as a "declaration reference".

But stepping back for a moment, I'm not sure I see any value in linking to generic parameter declarations, for the same reason that we don't link to regular parameters either.  They are very local things, and their documentation tends to be very small, so it's quite sufficient to link to the associated function/method/class instead.

@rbuckton Thus, it seems like we could completely eliminate `this._knownTypeParameters` logic by not generating a `canonicalReference` for `T` in the first place.

Reply:
The reason this is here is to prevent a possible collision with a type in scope that has the same name as a type parameter:

```ts
interface Foo {}
declare function f<Foo>(f: Foo): Foo;
```
In the output yaml, the `Foo` interface has `uid: !Foo:interface` and `name: Foo`. Unfortunately, I cannot recall whether docfx will attempt to resolve a `type: Foo` against a local `name:` if it can't resolve the uid, but I believe that was the reason why I added this in the first place. It's been more than a few weeks since I first looked at this.

Reply:
I can remove this logic. We can decide if we want to add it back in if we see issues with docfx.

Reply:
If I remember right, DocFX's file format has a design flaw where if a `type` value matches a UID, then it is assumed to be a reference to that UID.  They provide no escaping mechanism that could be used to represent a token that accidentally matches a UID, but is meant to be plain text (because it does not correspond to any linkable object).  If that's true, it is a flaw in the file format.

However, luckily your "canonical reference" notation will always contain a `!` character, which never appears in a valid type expression.  Thus, API Extractor should never encounter such a conflict.

It was possible type parameters only because the generic parameter is appearing in the `canonicalReference` field:

```js
{
  "kind": "Reference",
  "text": "T",
  "canonicalReference": "T"  // <=== not a correct reference
},
```